pkg/controller/v1alpha2/applicationconfiguration/render.go

```diff
@@ -494,12 +496,23 @@ func (r *components) getDataInput(ctx context.Context, s *dagSource, ac *unstruc
 	u.SetGroupVersionKind(obj.GroupVersionKind())
 	err := r.client.Get(ctx, key, u)
 	if err != nil {
-		reason := fmt.Sprintf("failed to get object (%s)", key.String())
+		reason := fmt.Sprintf("failed to get object (%s): %v", key.String(), err)
 		return nil, false, reason, errors.Wrap(resource.IgnoreNotFound(err), reason)
 	}
-	paved := fieldpath.Pave(u.UnstructuredContent())
+
 	pavedAC := fieldpath.Pave(ac.UnstructuredContent())
+	var acGeneration int
+	if err := pavedAC.GetValueInto("metadata.generation", &acGeneration); err != nil && !fieldpath.IsNotFound(err) {
+		return nil, false, err.Error(), err
+	}
 
+	// The source object's app generation should match current AC. Otherwise it is from an old AC and we should wait until
+	// it is updated then reconcile again.
+	if g1, g2 := u.GetLabels()[oam.LabelAppGeneration], strconv.Itoa(acGeneration); g1 != g2 {
+		return nil, false, fmt.Sprintf("generation not match: obj (%s), ac (%s)", g1, g2), nil
+	}
```

Comment:
The object can be workload or trait.

1. As a workload, the object will always have this label, it is meaningless to check.
2. As a trait, we have no way to guarantee the object has this label, in fact, no trait will have it.

And also, this is a break change, I don't think we need to make it as a force check.